From f0f4f66588944ae6c44bfe07b2fa8f771816ab1e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 7 May 2022 01:17:21 +0200
Subject: [PATCH 08/19] winex11.drv: Check for EWMH supported atoms during
 init.

CW-Bug-Id: #19553
CW-Bug-Id: #19644
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=2155

 100.0% dlls/winex11.drv/
diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 54e90140b0c..93f423d1a3e 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -3179,35 +3179,6 @@ LRESULT X11DRV_WindowMessage( HWND hwnd, UINT msg, WPARAM wp, LPARAM lp )
 }
 
 
-/***********************************************************************
- *              is_netwm_supported
- */
-static BOOL is_netwm_supported( Display *display, Atom atom )
-{
-    static Atom *net_supported;
-    static int net_supported_count = -1;
-    int i;
-
-    if (net_supported_count == -1)
-    {
-        Atom type;
-        int format;
-        unsigned long count, remaining;
-
-        if (!XGetWindowProperty( display, DefaultRootWindow(display), x11drv_atom(_NET_SUPPORTED), 0,
-                                 ~0UL, False, XA_ATOM, &type, &format, &count,
-                                 &remaining, (unsigned char **)&net_supported ))
-            net_supported_count = get_property_size( format, count ) / sizeof(Atom);
-        else
-            net_supported_count = 0;
-    }
-
-    for (i = 0; i < net_supported_count; i++)
-        if (net_supported[i] == atom) return TRUE;
-    return FALSE;
-}
-
-
 /***********************************************************************
  *              start_screensaver
  */
@@ -3286,7 +3257,7 @@ LRESULT X11DRV_SysCommand( HWND hwnd, WPARAM wparam, LPARAM lparam )
 
     if (NtUserGetWindowLongW( hwnd, GWL_STYLE ) & WS_MAXIMIZE) goto failed;
 
-    if (!is_netwm_supported( data->display, x11drv_atom(_NET_WM_MOVERESIZE) ))
+    if (!ewmh.has__net_wm_moveresize)
     {
         TRACE( "_NET_WM_MOVERESIZE not supported\n" );
         goto failed;
diff --git a/dlls/winex11.drv/x11drv.h b/dlls/winex11.drv/x11drv.h
index a10855805dc..e7633d16acb 100644
--- a/dlls/winex11.drv/x11drv.h
+++ b/dlls/winex11.drv/x11drv.h
@@ -436,6 +436,11 @@ static inline size_t get_property_size( int format, unsigned long count )
     return count * (format / 8);
 }
 
+struct x11drv_ewmh_data
+{
+    int has__net_wm_moveresize : 1;
+};
+
 extern XVisualInfo default_visual;
 extern XVisualInfo argb_visual;
 extern Colormap default_colormap;
@@ -462,6 +467,7 @@ extern int xrender_error_base;
 extern char *process_name;
 extern Display *clipboard_display;
 extern WNDPROC client_foreign_window_proc;
+extern struct x11drv_ewmh_data ewmh;
 
 /* atoms */
 
diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index 925e5fd7367..8f82ad7efde 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -88,6 +88,7 @@ int alloc_system_colors = 256;
 int xrender_error_base = 0;
 char *process_name = NULL;
 WNDPROC client_foreign_window_proc = NULL;
+struct x11drv_ewmh_data ewmh = {0};
 
 static x11drv_error_callback err_callback;   /* current callback for error */
 static Display *err_callback_display;        /* display callback is set for */
@@ -534,6 +535,37 @@ static void setup_options(void)
     NtClose( hkey );
 }
 
+
+/***********************************************************************
+ *              x11drv_ewmh_init
+ */
+static void x11drv_ewmh_init(void)
+{
+    Atom type, *supported;
+    unsigned long count, remaining;
+    char *atom_name;
+    int format, i, supported_count = 0;
+
+    if (!XGetWindowProperty( gdi_display, DefaultRootWindow(gdi_display), x11drv_atom(_NET_SUPPORTED), 0,
+                             ~0UL, False, XA_ATOM, &type, &format, &count, &remaining,
+                             (unsigned char **)&supported ))
+        supported_count = get_property_size( format, count ) / sizeof(Atom);
+
+    TRACE( "EWMH _NET_SUPPORTED:\n" );
+    for (i = 0; i < supported_count; ++i)
+    {
+        if (supported[i] == x11drv_atom(_NET_WM_MOVERESIZE))
+            ewmh.has__net_wm_moveresize = 1;
+
+        atom_name = XGetAtomName( gdi_display, supported[i] );
+        TRACE( "  %s\n", atom_name );
+        XFree( atom_name );
+    }
+
+    if (supported) XFree( supported );
+}
+
+
 #ifdef SONAME_LIBXCOMPOSITE
 
 #define MAKE_FUNCPTR(f) typeof(f) * p##f;
@@ -712,6 +744,7 @@ static NTSTATUS x11drv_init( void *arg )
     X11DRV_XComposite_Init();
 #endif
     x11drv_xinput2_load();
+    x11drv_ewmh_init();
 
     XkbUseExtension( gdi_display, NULL, NULL );
     X11DRV_InitKeyboard( gdi_display );
-- 
2.43.0

